cloudtools/start.py

```diff
@@ -161,4 +170,115 @@ def main(args):
+            "workerConfig": {
+                "numInstances": args.num_workers,
+                "machineTypeUri": args.worker_machine_type,
+                "diskConfig": {
+                    "bootDiskSizeGb": args.worker_boot_disk_size,
+                    "numLocalSsds": args.num_worker_local_ssds,
+                }
+            },
+            "softwareConfig": {
+                "imageVersion": image_version,
+                "properties": {
+                }
+            },
+            "initializationActions": []
+        }
+    }
+    
+    if args.num_preemptible_workers or args.preemptible_worker_boot_disk_size:
+        cluster_data['config']['secondaryWorkerConfig'] = {
+                "numInstances": args.num_preemptible_workers,
+                "machineTypeUri": args.worker_machine_type,
+                "diskConfig": {
+                    "bootDiskSizeGb": args.preemptible_worker_boot_disk_size,
+                    "numLocalSsds": args.num_worker_local_ssds,
+                },
+            }
+    
+    # add required metadata to the config json
+    metadata_json = cluster_data['config']['gceClusterConfig']['metadata']
+    metadata_json['HASH'] = hail_hash
+    metadata_json['SPARK'] = args.spark
+    metadata_json['HAIL_VERSION'] = args.version
+    # if we have user defined metadata, turn that into a dict and insert those keys into the metadata
+    if args.metadata:
+        for k,v in dict(s.split('=') for s in args.metadata.split(",")).iteritems():
+            metadata_json[k] = v
+    # if we have packages, add those to the metadata
+    if args.packages: 
+        metadata_json['PKGS'] = args.packages
+    
+    if args.jar:
+        metadata_json['JAR'] = args.jar
+    if args.zip:
+        metadata_json['ZIP'] = args.zip
+    
+    # convert the properties format to a dict that can be pushed into the cluster json
+    properties_json = cluster_data['config']['softwareConfig']['properties']
+    # note we join the properties and then split because properties before this point is a mix
+    # of a list of individual properties and the comma separated string that comes in from args
+    for k,v in dict(s.split('=') for s in ",".join(properties).split(",")).iteritems():
+        properties_json[k] = v
+    
+    # add each initialization action to the cluster json
+    for i in init_actions.split(","):
+        action = {"executableFile": i}
+        cluster_data['config']['initializationActions'].append(action)
+    
+    
+    dataproc = googleapiclient.discovery.build('dataproc', 'v1')
+    result = dataproc.projects().regions().clusters().create(
+        projectId=args.project,
+        region=args.region,
+        body=cluster_data).execute()
+    sys.stdout.write('\nWaiting for cluster creation...')
+    sys.stdout.flush()
+
+    
+    cluster_url = "https://console.cloud.google.com/dataproc/clusters/{}?project={}&region={})".format(args.name, args.project, args.region)
+    
+    while True:
```

Comment:
This is added to have it wait until the cluster is provisioned, but this step could be skipped with an optional flag if needed